man/man8/workqlat.8
.TH worklat 8  "2022-01-06" "USER COMMANDS"
.SH NAME
workqlat \- Work queue latency as a histogram.
.SH SYNOPSIS
.B workqlat [\-h] [\-W] [\-w func] [interval] [count]
.SH DESCRIPTION
This measures the time a work item spends waiting on a work queue after being
enqueued, and shows this time as a histogram.
This time should be small, but worker threads may not get scheduled quickly
enough due to CPU load causing work items to wait longer than usual.

This makes use of 2 tracing events, workqueue:workqueue_queue_work and
workqueue:workqueue_execute_start to record time when a work item was
enqueued and dequeued respectively. Since kernel can use same work object
with different handlers and also since same handler can be used with more
than one work objects, this tool uses a combination of work object and its
hanlder to identify each unique work.
This tool uses in-kernel eBPF maps for storing timestamps and the histogram,
for efficiency. Despite this, the overhead of this tool may become significant
for some workloads: see the OVERHEAD section.

It may need updating to match any changes to above mentioned tracing events.

Since this uses BPF, only the root user can use this tool.
.SH REQUIREMENTS
CONFIG_BPF and bcc.
.SH OPTIONS
.TP
\-h
Print usage message.
.TP
\-W
Show each work-handler item separately.
.TP
\-w func
Show work-handler named "func" only.
.TP
interval
Output interval, in seconds.
.TP
count
Number of outputs.
.SH EXAMPLES
.TP
Summarize work queue latency as a histogram:
#
.B workqlat
.TP
Print 5 second summaries, 3 times:
#
.B workqlat 5 3
.TP
Trace work handler "func" only, 1 second summaries:
#
.B workqlat -w func 1
.SH FIELDS
.TP
usecs
Microsecond range
.TP
count
How many times a task event fell into this range
.TP
distribution
An ASCII bar chart to visualize the distribution (count column)
.SH OVERHEAD
This traces 2 workqueue events which should not be too frequent to cause any
noticeable overhead for most workloads. While eBPF has very low overhead, and
this tool uses in-kernel maps for efficiency, the frequency of above mentioned
workqueue events for some workloads may be high enough that the overhead of this
tool becomes significant. Measure in a lab environment to quantify the overhead
before use.
.SH SOURCE
This is from bcc.
.IP
https://github.com/iovisor/bcc
.PP
Also look in the bcc distribution for a companion _examples.txt file containing
example usage, output, and commentary for this tool.
.SH OS
Linux
.SH STABILITY
Unstable - in development.
.SH AUTHOR
Imran Khan


debian/changelog
bcc (0.24.0-1) unstable; urgency=low

  * Support for kernel up to 5.16
  * bcc tools: update for trace.py, sslsniff.py, tcptop.py, hardirqs.py, etc.
  * new libbpf tools: bashreadline
  * allow specify wakeup_events for perf buffer
  * support BPF_MAP_TYPE_{INODE, TASK}_STORAGE maps
  * remove all deprecated libbpf function usage
  * remove P4/B language support
  * major test infra change, using github actions now
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 14 Jan 2022 17:00:00 +0000

bcc (0.23.0-1) unstable; urgency=low

  * Support for kernel up to 5.15
  * bcc tools: update for kvmexit.py, tcpv4connect.py, cachetop.py, cachestat.py, etc.
  * libbpf tools: update for update for mountsnoop, ksnoop, gethostlatency, etc.
  * fix renaming of task_struct->state
  * get pid namespace properly for a number of tools
  * initial work for more libbpf utilization (less section names)
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 15 Nov 2021 17:00:00 +0000

bcc (0.22.0-1) unstable; urgency=low

  * Support for kernel up to 5.14
  * add ipv4/ipv6 filter support for tcp trace tools
  * add python interface to attach raw perf events
  * fix tcpstates for incorrect display of dport
  * new options for bcc tools runqslower, argdist
  * new libbpf-tools: filetop, exitsnoop, tcprtt
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 15 Sep 2021 17:00:00 +0000

bcc (0.21.0-1) unstable; urgency=low

  * Support for kernel up to 5.13
  * support for debug information from libdebuginfod
  * finished support for map elements items_*_batch() APIs
  * add atomic_increment() API
  * support attach_func() and detach_func() in python
  * fix displaying PID instead of TID for many tools
  * new tools: kvmexit.py
  * new libbpf-tools: gethostlatency, statsnoop, fsdist and solisten
  * fix tools ttysnoop/readahead for newer kernels
  * doc update and bug fixes

 -- Yonghong Song <ys114321@gmail.com>  Mon, 16 Jul 2021 17:00:00 +0000

bcc (0.20.0-1) unstable; urgency=low

  * Support for kernel up to 5.12
  * Some basic support for MIPS
  * added bpf_map_lookup_batch and bpf_map_delete_batch support
  * tools/funclatency.py support nested or recursive functions
  * tools/biolatency.py can optionally print out average/total value
  * fix possible marco HAVE_BUILTIN_BSWAP redefine warning for kernel >= 5.10.
  * new tools: virtiostat
  * new libbpf-tools: ext4dist
  * doc update and bug fixes

 -- Yonghong Song <ys114321@gmail.com>  Mon, 5 May 2021 17:00:00 +0000

bcc (0.19.0-1) unstable; urgency=low

  * Support for kernel up to 5.11
  * allow BCC as a cmake subproject
  * add LPORT support in tcpconnlat and tcpconnect
  * added bpf_map_lookup_and_delete_batch support
  * new tools: virtiostat
  * new libbpf-tools: cpufreq, funclatency, cachestat
  * add install target to libbpf-tools
  * a few lua fixes
  * doc update and bug fixes

 -- Yonghong Song <ys114321@gmail.com>  Mon, 19 Mar 2021 17:00:00 +0000

bcc (0.18.0-1) unstable; urgency=low

  * Support for kernel up to 5.10
  * add bpf kfunc/kretfunc C++ example
  * add PT_REGS_PARMx_SYSCALL helper macro
  * biolatency: allow json output
  * biolatpcts: support measuring overall latencies between two events
  * fix build when ENABLE_CLANG_JIT is disabled
  * doc update and bug fixes

 -- Yonghong Song <ys114321@gmail.com>  Mon, 4 Jan 2021 17:00:00 +0000

bcc (0.17.0-1) unstable; urgency=low

  * Support for kernel up to 5.9
  * usdt: add uprobe refcnt support
  * use newer llvm/clang versions in debian packaging if possible
  * add bpf iterator C++ support
  * new bcc tools: tcprtt, netqtop, swapin, tcpsynbl, threadsnoop
  * tcpconnect: add DNS correlation to connect tracking
  * new libbpf-tools: llcstat, numamove, runqlen, runqlat, softirgs, hardirqs
  * doc update, bug fixes and some additional arguments for tools

 -- Yonghong Song <ys114321@gmail.com>  Thu, 29 Oct 2020 17:00:00 +0000

bcc (0.16.0-1) unstable; urgency=low

  * Support for kernel up to 5.8
  * trace.py: support kprobe/uprobe func offset
  * support raw perf config for perf_event_open in python
  * add BPFQueueStackTable support
  * added Ringbuf support support
  * libbpf-tools: readahead, biosnoop, bitesize, tcpconnlat, biopattern, biostacks
  * bug fixes and some additional arguments for tools

 -- Yonghong Song <ys114321@gmail.com>  Sat, 22 Aug 2020 17:00:00 +0000

bcc (0.15.0-1) unstable; urgency=low

  * Support for kernel up to 5.7
  * new tools: funcinterval.py, dirtop.py
  * support lsm bpf programs
  * support multiple pid/tids for offwaketime
  * usdt: add helpers to set semaphore values
  * turn off x86 jump table optimization during jit compilation
  * add support to use bpf_probe_read[_str_}{_user,kernel} in all bpf
  *    programs, fail back to old bpf_probe_read[_str] for old kernels
  * tools: add filtering by mount namespace
  * libbpf-tools: cpudist, syscount, execsnoop, vfsstat
  * lots of bug fixes and a few additional arguments for tools

 -- Yonghong Song <ys114321@gmail.com>  Mon, 19 Jun 2020 17:00:00 +0000

bcc (0.14.0-1) unstable; urgency=low

  * Support for kernel up to 5.6
  * new tools: biolatpcts.py
  * libbpf-tools: tools based on CORE and libbpf library directly
  * add --cgroupmap to various tools, filtering based cgroup
  * support kfunc (faster kprobe) for vfsstat, klockstat and opensnoop
  * lots of bug fixes and a few additional arguments for tools

 -- Yonghong Song <ys114321@gmail.com>  Mon, 20 Apr 2020 17:00:00 +0000

bcc (0.13.0-1) unstable; urgency=low

  * Support for kernel up to 5.5
  * bindsnoop tool to track tcp/udp bind information
  * added compile-once run-everywhere based libbpf-tools, currently
    only runqslower is implemented.
  * new map support: sockhash, sockmap, sk_storage, cgroup_storage
  * enable to run github actions on the diff

tools/workqlat.py
#!/usr/libexec/platform-python
# @lint-avoid-python-3-compatibility-imports
#
# workqlat   Work queue latency as a histogram.
#           For Linux, uses BCC, eBPF.
#
# USAGE: workqlat [-w] [interval] [count]
#
# This measures the time a work item spends waiting on a work queue before
# being executed by a worker, and shows this time as a histogram. This time
# should be small, but execution of work item may get delayed due to CPU load.
#
#
# Copyright 2022 Oracle and/or its affiliates.
# Licensed under the Apache License, Version 2.0 (the "License")
#
# Inspired by runqlat
# 06-Jan-2022   Imran Khan   Created this.

from __future__ import (
    absolute_import, division, unicode_literals, print_function
)
from bcc import BPF
from time import sleep
import argparse

# arguments
examples = """examples:
    ./workqlat           # summarize work queue latency as a histogram
    ./workqlat 1 10      # print 1 second summaries, 10 times
    ./workqlat -W        # show each work-handler item separately
    ./workqlat -w xyz    # show specified work handler
"""
parser = argparse.ArgumentParser(
    description="Summarize work queue latency as a histogram",
    formatter_class=argparse.RawDescriptionHelpFormatter,
    epilog=examples)
parser.add_argument("-W", "--works", action="store_true",
    help="print a histogram per work-handler item")
parser.add_argument("-w", "--work",
    help="trace only specified work item")
parser.add_argument("interval", nargs="?", default=99999999,
    help="output interval, in seconds")
parser.add_argument("count", nargs="?", default=99999999,
    help="number of outputs")
parser.add_argument("--ebpf", action="store_true",
    help=argparse.SUPPRESS)
arguments = parser.parse_args()
countdown = int(arguments.count)
debug = 0

src = """
#include <uapi/linux/ptrace.h>

typedef struct work_key {
    u64 work;
    u64 handle;
} work_key_t;

typedef struct work_handle_key {
    u64 handle;
    u64 slot;
} work_handle_key_t;

BPF_HASH(record, work_key_t);

STORAGE

TRACEPOINT_PROBE(workqueue, workqueue_queue_work) {
    work_key_t key = {};

    FILTER

    u64 ts = bpf_ktime_get_ns();
    key.work = (u64)args->work;
    key.handle = (u64)args->function;
    record.update(&key, &ts);

    return 0;
}

TRACEPOINT_PROBE(workqueue, workqueue_execute_start) {
    work_key_t key = {};
    u64 *ts = NULL;
    u64 delta = 0;
    u64 handle = 0;

    FILTER

    key.work = (u64)args->work;
    key.handle = (u64)args->function;
    ts = record.lookup(&key);
    if (ts != NULL) {
        delta = bpf_ktime_get_ns() - *ts;
        delta = delta / 1000; //nanosec to microsec
        handle = (u64)args->function;
        STORE
        record.delete(&key);
    }

    return 0;
}
"""

# code substitutions
label = "usecs"

if arguments.works:
    section = "work handles"
    handle = "handle"
    src = src.replace('STORAGE', 'BPF_HISTOGRAM(dist, work_handle_key_t);')
    src = src.replace('STORE',
            'work_handle_key_t hkey = {.handle = ' + handle
            + ', .slot = bpf_log2l(delta)}; '
            + 'dist.increment(hkey);')
else:
    section = ""
    src = src.replace('STORAGE', 'BPF_HISTOGRAM(dist);')
    src = src.replace('STORE', 'dist.increment(bpf_log2l(delta));')

if arguments.work:
    func_name = BPF.ksymname((arguments.work))
    if (func_name < 0):
        print("Invalid work handler \n")
        exit()

    src = src.replace('FILTER',
        """u64 val = 0x%x;
        int ret;
        if ((u64)args->function != val)
        return 0;""" % func_name)

else:
    src = src.replace('FILTER', '')

if debug or arguments.ebpf:
    print(bpf_text)
    if arguments.ebpf:
        exit()

# load BPF program
b = BPF(text=src)

if arguments.work:
    print("Tracing work queue latency for "
          + arguments.work + "... Hit Ctrl-C to end.")
else:
    print("Tracing work queue latency ... Hit Ctrl-C to end.")

# output
exiting = 0 if arguments.interval else 1
dist = b.get_table("dist")
while (1):
    try:
        sleep(int(arguments.interval))
    except KeyboardInterrupt:
        exiting = 1

    print()
    print("Printing histogram")
    dist.print_log2_hist(label, section, section_print_fn=b.ksym)
    dist.clear()

    countdown -= 1
    if exiting or countdown == 0:
        exit()
